servlet/src/main/scala/org/http4s/servlet/AsyncHttp4sServlet.scala

```diff
@@ -57,22 +58,20 @@ class AsyncHttp4sServlet[F[_]](
   private def handleRequest(
       ctx: AsyncContext,
       request: Request[F],
-      bodyWriter: BodyWriter[F]): F[Unit] = {
-    val timeout = F.async[Unit] { cb =>
-      ctx.addListener(new AsyncTimeoutHandler(cb))
-    }
+      bodyWriter: BodyWriter[F]): F[Unit] = Deferred[F, Unit].flatMap { gate =>
+    // It is an error to add a listener to an async context that is
+    // already completed, so we must take care to add the listener
+    // before the response can complete.
+    val timeout =
+      F.asyncF[Response[F]](cb => gate.complete(ctx.addListener(new AsyncTimeoutHandler(cb))))
     val response =
-      optionTSync
-        .suspend(serviceFn(request))
-        .getOrElse(Response.notFound)
-        .recoverWith(serviceErrorHandler(request))
+      gate.get *>
```

Comment:
Sorry, maybe I'm missing something, but I don't see difference between race with synchronization on gate  and plain flatMap i.e. `timeout *> response` because we will wait `gate.get` completion

Reply:
This makes `response` wait on the call to `ctx.addListener`, not the asynchronous completion of `timeout`.  The original bug was because `response` could complete `ctx` before `ctx.addListener` was called.  And `timeout *> response` would wait for the timeout to fire before attempting `response`.

Reply:
Sorry, my brain have completely ignored part with `F.asyncF[Response[F]](cb =>...)`, and now with your explanation i've finally understood whole picture. Thanks for explanation!

Reply:
No problem.  Thanks for the bug report!  This should be released in 0.19.0-M3 as soon as the fs2 release candidate is out.